### PR TITLE
Claim back unused invites #328

### LINKED ISSUE #328: Claim back unused invites
We will send unused invites back to the user after 2 weeks. 

Saves memory on the contracts and cleans up the clutter

## Spec

- Scheduler calls into join.seeds "chkcleanup"

- Join.seeds has a new table called timestamp_table and a function cleanup(from_id, to_id) that cleans unclaimed invites between 2 indices. (already implemented)

timestamp table has
id
invite_id
timestamp

invite_id is the most recent invite id currently in the invites table
timestamp is the current time

chkcleanup does the following
check if timestamp table is empty - 

if not empty, initiate a cleanup action on all IDs lower than that ID

whether or not it's empty, add a new entry to the timestamp table with the current last invite ID, and the current timestamp

have some safety check that timestamp isn't < 3 days old? or enforce setting time interval between cleanups? 

One of these 2 - we normally do not trust the scheduler to be accurate with their calls, so should enforce the interval here, and the scheduler can then run every day.

### PR DESCRIPTION
**Pull request by IanMendozaJaimes**
- Added `chkcleanup` action to onboarding and added to the scheduler
- Added tests for onboarding and refactored scheduler tests

Fixes #328 

Contracts that need to be deployed:
- [ ] Onboarding
- [ ] Scheduler

Other considerations:
- [ ] Permissions need to be updated
- [ ] Scheduler needs to be updated

### CODE REVIEW

File: src/seeds.onboarding.cpp [hunk 1]
```diff
@@ -389,11 +395,62 @@ void onboarding::accept(name account, checksum256 invite_secret, string publicKe
...
+  timestamp_tables timestamps(get_self(), get_self().value);
+  invite_tables invites(get_self(), get_self().value);
+
+  if (invites.begin() == invites.end()) { return; }
+
+  if (timestamps.begin() != timestamps.end()) {
+
+    auto titr = timestamps.end();
+    titr--;
+
```
Comment (n13) [inline src/seeds.onboarding.cpp:410]: Not sure this works, IMO doing "rbegin()" is better than end() -- 1

File: src/seeds.onboarding.cpp [hunk 2]
```diff
@@ -389,11 +395,62 @@ void onboarding::accept(name account, checksum256 invite_secret, string publicKe
...
+      get_self(),
+      "cleanup"_n,
+      std::make_tuple(start_id, max_id, config_get("batchsize"_n))
+    ).send();
+
+  }
+
+  auto last_invite_itr = invites.end();
+  last_invite_itr--;
+
```
Comment (n13) [inline src/seeds.onboarding.cpp:440]: same here - use rbegin()

File: src/seeds.onboarding.cpp [hunk 3]
```diff
@@ -389,11 +395,62 @@ void onboarding::accept(name account, checksum256 invite_secret, string publicKe
...
+      if (prev_itr->invite_id < max_id) {
+        start_id = prev_itr->invite_id;
+      }
+    }
+
+    action(
+      permission_level(get_self(), "active"_n),
+      get_self(),
+      "cleanup"_n,
+      std::make_tuple(start_id, max_id, config_get("batchsize"_n))
```
Comment (n13) [inline src/seeds.onboarding.cpp:433]: not sure cleanup can handle 200...

Comment (n13) [reply src/seeds.onboarding.cpp:433]: actually should be fine, it's x8 for each returned invite... 25 returned invites hopefully works

**Review by n13 — CHANGES_REQUESTED** [review]
Looks perfect, only one minor change requested

**Comment by IanMendozaJaimes** [comment]
- I have changed the iterators in the chkcleanup function

**Review by n13 — APPROVED** [review]
Looks good! Will deploy shortly!

**Comment by n13** [comment]
I merged, but can't get onboarding test to succeed, can you look at it? Can't deploy until it's passing

'''
      next_key: ''
    }    ✔  Given cleanup ran: should have the correct timestamp entries

    {
      rows: [
        {
          invite_id: 0,
          transfer_quantity: '10.0000 SEEDS',
          sow_quantity: '5.0000 SEEDS',
          sponsor: 'seedsuseraaa',
          account: '',
          invite_hash: '063e5deabe176d0bc2ba72724f8f0696319ddeed7800b5109751f26fd0f02e51',
          invite_secret: '0000000000000000000000000000000000000000000000000000000000000000'
        },
        {
          invite_id: 1,
          transfer_quantity: '10.0000 SEEDS',
          sow_quantity: '5.0000 SEEDS',
          sponsor: 'seedsuserbbb',
          account: '',
          invite_hash: 'af6d775fcd1110329bc3a7e058171d8c5502436fe96624d24fe79c22a91efeb5',
          invite_secret: '0000000000000000000000000000000000000000000000000000000000000000'
        },
        {
          invite_id: 2,
          transfer_quantity: '10.0000 SEEDS',
          sow_quantity: '5.0000 SEEDS',
          sponsor: 'seedsuserbbb',
          account: '',
          invite_hash: 'b715fe2cb69014f9b7dddfabe8449924cdd787b595003180d72cff98bb3f5b57',
          invite_secret: '0000000000000000000000000000000000000000000000000000000000000000'
        }
      ],
      more: false,
      next_key: ''
    }    ✖  Given cleanup ran: should have the correct number of invites
    ----------------------------------------------------------------
        operator: deepEqual
        diff: "1" => "3"
        source: at checkInvites (/Users/elohim/play/seeds/seeds-smart-contracts/test/onboarding.test.js:1071:9)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async /Users/elohim/play/seeds/seeds-smart-contracts/test/onboarding.test.js:1116:5
        
        

    {
      rows: [ { id: 0, invite_id: 2, timestamp: 1623584299 } ],
      more: false,
      next_key: ''
    }    ✖  Given cleanup ran: should have the correct timestamp entries
    ----------------------------------------------------------------
        operator: deepEqual
        expected: [{"id": 0, "invite_id": 2}, {"id": 1, "invite_id": 2}]
        diff: [
        - 1: {"id":1,"invite_id":2}
        ]
        source: at checkInvites (/Users/elohim/play/seeds/seeds-smart-contracts/test/onboarding.test.js:1086:9)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async /Users/elohim/play/seeds/seeds-smart-contracts/test/onboarding.test.js:1116:5
        
        



  passed: 41,  failed: 2  of 43 tests  (15.1s)

'''

**Comment by n13** [comment]
Un-merged, but status didn't update

**Comment by n13** [comment]
Deployed as new PR - see here https://github.com/JoinSEEDS/seeds-smart-contracts/pull/360

**Pull request merged**